### PR TITLE
[WIP] tensornetwork torch backend

### PR DESCRIPTION
**Pull request by co9olguy**
**Context:** Adding support for the tensornetwork backend to directly execute the entire computation in pytorch

**Description of the Change:** This is a WIP. So far I have just hacked the pipeline to see where we need to change things. Some things are simple fixes (checking for torch tensor types), others are bigger blockers (torch does not support complex number tensors).

**Benefits:** we need to do this if we want to enable backpropagation for gradient computations (can be much faster than parameter-shift rule)

**Possible Drawbacks:** Right now, the biggest blocker seems to be lack of complex number support, as well as potentially needing to make pytorch an explicit dependency (since there are several spots along the pipeline where we need to check if something is a ``torch.Tensor``)

**Related GitHub Issues:** N/A

===================================================

Steps needed to ensure the whole pipeline worked:

- ``args_to_numpy`` not really needed (only to convert tuple->list)
- ``_flatten`` and ``unflatten`` functions need to be made compatible with ``torch.Tensor``s
- ``Variable.free_param_values = list(_flatten(args))`` casts to numpy array (lose torch.Tensors). This happens twice in qnode.py. Probably ok for kwargs?
- ``self.output_conversion = float`` needs to keep Tensors as Tensors
- ``if not isinstance(p, numbers.Real)`` in operation.py does not recognize a real-valued torch.Tensors
- will need to manually code up backend-specific (or dispatched) versions of all matrix-creating functions for tensornetwork
- pytorch doesn't like complex tensors :(
- need to declare backend when creating tn.Node (it doesn't infer from data)
- torch doesn't seem to like tensordot when the arguments contain both numpy arrays and tensors
- need to convert observables to tensors before making them into nodes
- pytorch tensordot complains if the two tensors have different numeric dtypes (float/double)
- ``expval.imag`` and ``expval.real`` in tensornetwork ``ev`` method fail for pytorch tensors
- ``execute`` has a final line ``return np.asarray(results)``, which kills the tensor
- ``self.output_conversion(ret)`` in qnode.py does not work because it references the function `float`
- ``forward`` attempts to convert the result of the qnode to a numpy array
- ``return torch.from_numpy(res)`` in `_TorchQNode` attempts to create a torch object from numpy (but if the object is already a torch tensor, it fails)

Other notes:

- There still seems to be a lot of autograd stuff hanging around (i.e., a torch qnode is still wrapped by autograd as an ``ae.primative``). Perhaps this will be fixed by ``JacobianQNode`` PR
- we do a lot of flattening/unflattening multiple times. Might it make sense to just do this once at start/finish

### CODE REVIEW

**Comment by josh146** [comment]
Closing this for now, as this PR is outdated, and the path to creating a Torch backend looks somewhat different.
